lib/DBD/mysql.pm

```diff
@@ -1186,6 +1186,18 @@ flags:
 These are used to specify the respective parameters of a call
 to mysql_ssl_set, if mysql_ssl is turned on.
 
+=item mysql_ssl_verify_server_cert
+
+This turns on the MYSQL_OPT_SSL_VERIFY_SERVER_CERT connection option.
+
+Note that this is deprecated as of MySQL 5.7.11 and is removed in MySQL 8.0
+
+=item mysql_ssl_mode
+
+Can be set to 'disabled', 'preferred', 'required', 'verify_ca' or 'verify_identity'.
+Default is 'preferred'.
+
+Available from MySQL 5.7.11
```

Comment:
I started to work on this because I wanted to use `SSL_MODE_VERIFY_CA`, and that wasn't possible in `DBD::mysql`. At that time I knew that Oracle MySQL was moving from `--ssl` and `--ssl-verify-server-cert` to `--ssl-mode`.
* [WL#8785: Support multi-state SSL option](https://dev.mysql.com/worklog/task/?id=8785)
* [WL#9091: Remove --ssl, --ssl-verify-server-cert client-side options](https://dev.mysql.com/worklog/task/?id=9091)

I sort of expected MariaDB to follow the same path. But I now know that MariaDB server will most likely not follow that path and keep using the 'old' variables.

So I think you're right and `DBD::mysql` should implement a stable API.

DBD::mysql | MySQL 5.7+ | MariaDB
----------------|-----------------|-------------
**ssl=0**, ssl-verify-server-cert=0 (default) | `SSL_MODE_DISABLED` | ssl=0, ssl-verify-server-cert=0
ssl=?, **ssl_prefer=1** (new), ssl-verify-server-cert=0 | `SSL_MODE_PREFERRED` | Not set anything and let it use the defaults? 
**ssl=1**, ssl-verify-server-cert=0 | `SSL_MODE_REQUIRED` | ssl=1
ssl=1, ssl-verify-server-cert=0, **mysql_ssl_ca_(file_path)=xxx** | `SSL_MODE_VERIFY_CA` | set ca_file/ca_path for `mysql_ssl_set()`
ssl=1, **ssl-verify-server-cert=1** | `SSL_MODE_VERIFY_IDENTITY` | `MYSQL_OPT_SSL_VERIFY_SERVER_CERT`

`SSL_MODE_VERIFY_CA` is in many cases the highest level which still works. The certificates which are [generated by default](https://dev.mysql.com/doc/refman/5.7/en/creating-ssl-rsa-files-using-mysql.html) have a CN which is similar to "MySQL_Server_5.7.16_Auto_Generated_CA_Certificate". And even if one would use custom certificates with a hostname then it might not be possible to enable hostname validation because of the [missing support for SubjectAltName](https://bugs.mysql.com/bug.php?id=68052) (also for [MariaDB](https://jira.mariadb.org/browse/MDEV-10594) ). So using a virtual-IP, loadbalancer etc. can make it impossible to enable this.

Another question is what the default should be. In Oracle MySQL the default seems to be 'preferred' (see https://dev.mysql.com/doc/refman/5.7/en/secure-connection-options.html)
I do think the default should eventually be set to `SSL_MODE_VERIFY_IDENTITY`/`verify-server-cert=1,ssl=1`. Similar to what browsers and `curl` does by default. 

So I would suggest to do this:
1. Add `ssl_prefer`. Keep disabled by default for backwards compatibility. Maybe change this to `ssl_prefer=1` at a later stage?
2. Set `SSL_MODE_VERIFY_CA` if a ca file is specified (probably already what is happening now, but this could make it more explicit.

What should we expect to happen with MariaDB and `mysql_ssl_preferred`?

With `DBD::mysql` and MySQL 5.6:
[`--ssl`](https://dev.mysql.com/doc/refman/5.6/en/secure-connection-options.html#option_general_ssl) was similar to `SSL_MODE_PREFERRED`, this was changed in MySQL 5.7.6 to be `SSL_MODE_REQUIRED`. So compiling against 5.7 might suddenly cause a different behaviour. In 5.6 and older there was no real option to force the use of SSL/TLS on the client.